plugins/index.ts

```diff
@@ -179,23 +179,44 @@ extendEnvironment((hre) => {
       hre.log(
         'Found existing proxy at:',
         proxyAddress,
-        ' attempting to upgrade instance',
+        'attempting to upgrade instance',
         contractName
       );
+      const existingImplementationAddress =
+        await hre.upgrades.erc1967.getImplementationAddress(proxyAddress!);
+      hre.log('Existing implementation at:', existingImplementationAddress);
       const fireblocksSigner = signer as FireblocksSigner;
       if (typeof fireblocksSigner.setNextTransactionMemo === 'function') {
-          fireblocksSigner.setNextTransactionMemo(`Upgrade contract instance for ${contractName}`);
+        fireblocksSigner.setNextTransactionMemo(
+          `Upgrade contract instance for ${contractName}`
+        );
+      }
+      const deployment = await hre.deployments.get(contractName);
+      const artifact = await hre.deployments.getArtifact(contractName);
+      if (deployment.bytecode !== artifact.bytecode) {
+        contract = await hre.upgrades.upgradeProxy<TContract>(
+          proxyAddress!,
+          contractFactory
+          // options
+        );
+        const newImplementationAddress =
+          await hre.upgrades.erc1967.getImplementationAddress(proxyAddress!);
+        if (existingImplementationAddress === newImplementationAddress) {
+          hre.log('Implementation unchanged');
+        } else {
+          hre.log('New implementation at:', newImplementationAddress);
+        }
+        hre.trace('...awaiting deployment transaction', contractName);
+        await contract.deployed();
+        hre.log('...successful deployment transaction', contractName);
+      } else {
+        hre.log('Implementation appears unchanged, skipped upgrade attempt.');
+        contract = (await getContract({
+          contractName,
+          hre,
+        })) as InstanceOfContract<TContract>;
       }
-      contract = await hre.upgrades.upgradeProxy<TContract>(
-        proxyAddress,
-        contractFactory
-        // options
-      );
-      hre.log('Upgraded instance', contractName, 'at', contract.address);
     }
-    hre.trace('...awaiting deployment transaction', contractName);
-    await contract.deployed();
```

Comment:
why did this get removed?

Reply:
I've seen the absence of this (as well as tx.wait) cause race conditions in deploy code, test code, etc

Reply:
Hmm, good question, I will put that back in.